API/ERR: allow iterators in df.set_index & improve errors

h-vetinari opened the pull request:
- [x] closes the parts of #22484 (resp. those worth keeping) that were reverted in #25085 due to #24969 ~closes #24969~
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

This is a quick fix for the regression - however, I think this should be immediately (i.e. 0.24.1) be deprecated. I haven't yet added a deprecation warning here, pending further discussion in the issue.

@jorisvandenbossche @TomAugspurger @jreback

h-vetinari committed:
API: re-enable custom label types in set_index

h-vetinari commented on pandas/core/frame.py:
```diff
@@ -4135,13 +4135,8 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
-            # make sure we have a container of keys/arrays we can iterate over
-            # tuples can appear as valid column keys!
+        if not isinstance(keys, list):
             keys = [keys]
```
If supporting custom types, we need to go back to the state pre-#22486 that just puts everything that's not listlike in list. Otherwise we can't guarantee that iteration below will work.

jreback replied:
I think that is best, let's just revert entirely to pre 22486

h-vetinari replied:
pre-#22486 had some other problems we do not need to re-instate (e.g. weird KeyErrors if it's an iter).

I agree that the code complexity must stay maintainable, and so the most reasonable thing (IMO), ist to just not bother trying to "fix" **unhashable** custom types being used as keys. At that point, the user is so far gone off the reservation that it's really not our job to give them a good error message (again, even the repr of such a frame would crash).

wkschwartz commented on pandas/core/frame.py:
```diff
@@ -4150,10 +4145,20 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
+                # everything else gets tried as a key; see GH 24969
+                try:
+                    self[col]
+                    str(col)
```
Can `str(col)` normally raise a KeyError?

h-vetinari replied:
The error might be different, that's true. On second thought, `str` defaults to `repr` if it's not implemented, so unless the class manages to break its own `repr` (haha), this should not be necessary.

h-vetinari committed:
Fix doc pattern?

TomAugspurger commented on pandas/tests/frame/test_alter_axes.py:
```diff
@@ -270,6 +270,31 @@ def test_set_index_raise_on_type(self, frame_of_index_cols, box,
+                self.name = name
+                self.color = color
+
+            def __str__(self):
```
Is `__str__` used still?

h-vetinari replied:
Would work without, just means the repr is ugly (and maybe harder to debug if someone doesn't know what's happening):
```
>>> df = pd.DataFrame({thing1: [0, 1], thing2: [2, 3]})
>>> df
   <__main__.Thing object at 0x000001973BAB9908>  <__main__.Thing object at 0x000001973BAB99B0>
0                                              0                                              2
1                                              1                                              3
```

TomAugspurger replied:
👍. Just (overly) concerned about behavior differences between objects where `__str__` is / isn't defined. Best to include it.

TomAugspurger reviewed (commented):
Thanks for the PR.

I'm not sure what should be done going forward, or whether we should have a deprecation now. @toobaz may have thoughts here too. I'm fine with leaving that for a comprehensive "what's allowed an Index" overhaul.

wkschwartz commented:
Reposting https://github.com/pandas-dev/pandas/issues/24969#issuecomment-458246305 here in case the discussion of the deprecation of custom label types proceeds here instead of at the original ticket.

> Obviously I would prefer no deprecation of custom label types as they are integral to my company’s applications. However, I would urge strongly that if you do decide to deprecate the feature, you do so starting only in the next major release (presumably 0.25.0) rather than in a minor release (0.24.1).
>
> If you do stop supporting custom label types and I am not to be stuck at Pandas 0.23.4 forever, I could theoretically undertake the (expensive) refactoring to use unique IDs (my production code has the equivalent of the name field from my toy example in the OP). However, other users whose code would break might not have convenient unique IDs to switch to. Please do not remove this feature lightly.

toobaz commented:
> @toobaz may have thoughts here too

I'm boring, I always have the same thoughts, every time the discussion comes out :-) And they are:
- if we didn't say it's not allowed, it's allowed, and should work
- if there is no reason to disallow it, let's not say it's not allowed
- if we think that the reason to disallow it is that it makes code simpler, we are almost always wrong. Clean code will work with any type a user throws in, as long as it satisfies general properties (e.g. ``__hash__``). Viceversa, explicitly listing the types we "like" as scalar will require us to amend the list (and the docs) every time we find some new type that yes, we think is worth allowing.
- (a _valid_ reason to disallow it is that either it brings ambiguity, or it changes the semantics, as would be accepting lists as keys)

toobaz commented:
Don't get me wrong: it is good to have this discussion once and for all if it results in cleaner docs/assumptions. But my opinion is that the docs should say "you can use as keys anything which is not mutable".

toobaz commented:
> "Not mutable" is a bit fuzzy in Python. If, instead, keys are required to be _hashable_

Yep, my mistake, "hashability" is the right property.

jreback commented on pandas/tests/frame/test_alter_axes.py:
```diff
@@ -270,6 +270,31 @@ def test_set_index_raise_on_type(self, frame_of_index_cols, box,
+        thing2 = Thing('Two', 'blue')
+        df = DataFrame({thing1: [0, 1], thing2: [2, 3]})
+        expected = DataFrame({thing1: [0, 1]},
+                             index=Index([2, 3], name=thing2))
```
I guess can you add an object that is not hashable (maybe just inherit an object from list) and see what happens

h-vetinari replied:
To test what? Failure?
Lists would currently be interpreted as array-likes.

I am already testing sets, which are not hashable.

jreback replied:
no thus is a test of a custom class that fails hashing for example

jorisvandenbossche replied:
But you cannot create a dataframe with an unhashable object as column name in the first place, so you would actually be testing that you are passing an object that is not in the column names, which is already covered I think?

TomAugspurger replied:
That's what I thought, but...

```python
df = pd.DataFrame([[0, 1]], columns=[{'a'}, {'b'}])
```

though essentially everything is broken on this dataframe :).

jorisvandenbossche replied:
So then I would say we shouldn't be testing that, but rather open an issue about it ? :-)

Although I suppose we simply don't test hashability of object-dtype ndarrays at Index creation time, but only when it is used. Which makes sense, as I suppose it can be quite costly to do so.

h-vetinari replied:
Adding a custom class that is unhashable is such a pathological edge case that it really does not make sense to add a test. As @TomAugspurger showed, the constructor essentially goes through, but then almost everything else is broken, even the repr.

Unhashable types (in this case `set` and `iter`) are already tested to raise an error.

h-vetinari replied:
Ok, I found a way to still reliably test this. But then the question if `set` should raise a `ValueError` as currently, or a `TypeError` as something that's not hashable. I went with the latter, as it would be the only consistent treatment for what has been specified by @jreback above (i.e. `TypeError` for unhashable)

h-vetinari commented:
I'm not a core dev, but FWIW, here are some counter points to @toobaz arguments:

> if there is no reason to disallow it, let's not say it's not allowed

If the code is not designed and *tested* for it, such a carte blanche is (not in this example, but across 1000s of small cuts) an immense burden down the road when something that wasn't explicitly forbidden suddenly breaks.

> if we didn't say it's not allowed, it's allowed, and should work

See lack of testing above. Duck typing really bites us in the ass here. Stability (a possible SemVer after 1.0) even more.

> if we think that the reason to disallow it is that it makes code simpler, we are almost always wrong. Clean code will work with any type a user throws in, as long as it satisfies general properties (e.g. `__hash__`).

The previous behaviour was specifically causing some weird KeyErrors for stuff that should have never been tested as a key (hence I'm adding a try-except in this PR), see #22484. I could have easily maintained this capability, if it had been tested, or even documented.

> Viceversa, explicitly listing the types we "like" as scalar will require us to amend the list (and the docs) every time we find some new type that yes, we think is worth allowing.

That - in contrast to the above - would make it a controlled (non-breaking) expansion where necessary (IMO far preferable). I'm also thinking the ecosystem has matured enough by now to broadly have a common understanding what "scalar" means.

> (a valid reason to disallow it is that either it brings ambiguity, or it changes the semantics, as would be accepting lists as keys)

That line can be blurred arbitrarily far (see [my example](https://github.com/pandas-dev/pandas/issues/24969#issuecomment-458241015) in #24969).

> Don't get me wrong: it is good to have this discussion once and for all if it results in cleaner docs/assumptions. But my opinion is that the docs should say "you can use as keys anything which is not mutable".

In #24702, it sounded like you're advocating against using tuples as keys (which I would agree with), which [also aren't mutable](https://docs.python.org/3/glossary.html#term-immutable)

IMO, the API should have a clearly defined surface, not an arbitrary "whatever happens to work". It would also mean (and already does to some extent) a disproportionate amount of dev time to chase down weird corner cases, rather than focus on a consistent API that's useful and unambiguous for 99.99% of the cases.

h-vetinari commented:
> You test lines of code, not user behavior. Otherwise tests suites would have to be infinite.

Not if there's a well-defined set of input types. ;-)

> Feel free to expand, I'm not following you.

The iterator that gets passed to `df.set_index` (see OP of #22484), say, in the expectation of its values being used as the index, somehow gets turned into an `Index` and then raises a `KeyError`. One of several examples.

> That's sure. I'm being very clear (although @wkschwartz beat me at it) on what the API should support. And it should work on what it supports.

Ok, that's a more nuanced discussion then (vs. "everything that works is allowed"). If all hashable objects should be usable as keys, we'd have to drastically improve our testing for it, but it's not impossible, certainly. I just think it's far too wide a line to draw. I mentioned in the issue that it would be more feasible to add `dataclasses` specifically to the types we test, rather than all hashables (and to be sure: my point is not about @wkschwartz use case - much less against him).

toobaz commented:
> If all hashable objects should be usable as keys, we'd have to drastically improve our testing for it, but it's not impossible

Again, non sequitur. Testing is good, but we want to write code that works even in those cases in which it was not tested.

> Not if there's a well-defined set of input types. ;-)

OK, let's cut it short: will you soon be asking us to limit the _values_ of ``Series``/``DataFrame``s to well-defined set of input types? I hope not. Do you think this is a problem for our testing suite? I hope not.

If you need to test any possible combination of allowed input types with every possible behavior of every method in the pandas API, tests will be just ''slightly short'' of infinite.


> (see OP of #22484)

I see three points there, and not sure if any of them is a problem. Can you assume I'm even more stupid than I am?


General comment: I still wasn't able to understand if this regression was a mistake or was made on purpose. In the second case, while I can only blame myself for not being able to keep up with the PRs, I still think such a change deserved some more discussion, e.g. in mailing list.

wkschwartz commented:
To expand on @toobaz's https://github.com/pandas-dev/pandas/pull/24984#issuecomment-458342687: in a duck-typed language, APIs should be based on protocols, interfaces, or capabilities, not on exact class hierarchies. For example, Python core does fine defining the API surface of `dict` keys as anything hashable.

Regarding whether [that which is not forbidden is allowed](https://en.wikipedia.org/wiki/Everything_which_is_not_forbidden_is_allowed) (TWINFIA): Pandas' having a version < 1 means the maintainers have the right to change semantics. Please balance that right against the mass of established code bases in production that rely on Pandas. I couldn't be the only author to assume TWINFIA.

Thank you, @h-vetinari and @toobaz, for taking an interest in this!

h-vetinari commented:
> Again, non sequitur. Testing is good, but we want to write code that works even in those cases in which it was not tested.

I agree that properly written code, that only ever depends on hashability for keys, is possible. But many methods (not just `set_index`) need to do a lot of gynmastics for **wildly varying input signatures** - in this case, column keys, list of column keys, various arrays, lists of arrays, and mixes between arrays and keys.
*This* to me is a large reason of the utility of pandas, because very different (but equally sensible) scenarios just work. Allowing arbitrary types for keys makes this "magic" much more complicated, but OK, opinions can differ what should be the design goal.

> > Not if there's a well-defined set of input types. ;-)

> OK, let's cut it short: [strawman]

Let me clarify my language here: it's certainly possible to test methods against various *container* types that it should accept. What's in those containers is another story.

> > (see OP of #22484)

> I see three points there, and not sure if any of them is a problem. Can you assume I'm even more stupid than I am?

I know you're not (stupid), but ok, here's another try:
```
>>> import pandas as pd
>>> import numpy as np
>>>
>>> df = pd.DataFrame(np.random.randn(5, 5), columns=list('ABCDE'))
>>>
>>> # Series-constructor uses iterators as if it was an array
>>> my_iter = map(str, df.A)
>>> pd.Series(my_iter)
0     0.8812414814169731
1    -1.0789948340108944
2    -0.2049936946205116
3    0.35952380472779194
4     -0.529524992759882
>>>
>>> # so why not do the same for df.set_index, which *also* works with arrays?
>>> my_iter = map(str, df.A)  # had been consumed previously
>>> df.set_index(my_iter)
Traceback (most recent call last):
[...]
KeyError: "None of [Index(['0.8812414814169731', '-1.0789948340108944', '-0.2049936946205116',\n       '0.35952380472779194', '-0.529524992759882'],\n      dtype='object')] are in the [columns]"
```
There's a couple things that are confusing here. How did an iterator get turned into an `Index` object, without any further user-input? And why is that `Index` being tested as a column key?!

> General comment: I still wasn't able to understand if this regression was a mistake or was made on purpose.

There was no informed decision on this regression because it was neither tested nor documented (if the documented standard had been "all hashable objects can be key", it would have been easy to conform to that). The problems were much more mundane, i.e. having to inspect the items of the outer-most container and determine whether they are keys / `Index` / `Series` / `MultiIndex` / `np.ndarray`. Through several rounds of review @jreback then added the review requirment to add objects that pass `is_list_like` (which got reverted just before the release) but that would have still broken this use case here.

The common ground here is (as we've seen above) to have a well-defined API. If what can be keys is explicitly documented somewhere, it can be kept in mind for whatever code that's being written, and hashability would be one of several valid choices for that.

My main point above was that enforcing code to stay runnable that only works because of an oversight (and there's a bunch of that in pandas, no matter how much we strive for clean code) is way too generous.

toobaz commented:
> But many methods (not just `set_index`) need to do a lot of gynmastics for **wildly varying input signatures**

Claim: identifying _lists_ (more in general, containers) requres gymnastics (and, incidentally, detailed docs), while identifying _valid keys_ is extremely simple if their definition is "hashable object". And requires more complicated code, more complicated docs, and more waste of the user's memory, if the definition is more complicated than that.

As in: ExtensionArray (and having even our own types rely on them) is already making (I think) the pandas codebase _cleaner_, and _easier_ to maintain, precisely because we gave up setting the admissible values.

> I know you're not (stupid), but ok, here's another try:

Thanks, I appreciate. You are perfectly right that pandas' definition of "list-like" is a mess, and that's a ground where we desperately need to set standards. Or maybe, we already all agree on standards (iterators _are_ list-like, tuples are not), and there is just annoying legacy code that doesn't follow them (for sure ``is_list_like`` [doesn't](https://github.com/pandas-dev/pandas/issues/24702)). I think that's not the case for keys.

TomAugspurger commented:
Steering the discussion back to the PR: what needs to be done for 0.24.1? IMO, the changes and tests look good, just need a release note under "fixed regressions". Anything else?

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -4150,10 +4145,19 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
+            else:
+                # everything else gets tried as a key; see GH 24969
+                try:
+                    self[col]
```
I think doing `col in self.columns` is a bit more efficient? (doesn't need construct the series) 
Although then you also need to care about the False case

h-vetinari replied:
I tried, but `col in self.columns` works without the hashing, and even if check
```
hash(col) and col in self.columns
```
this does not catch the `iter` case (an iterator is hashable, who would have thought...)

jorisvandenbossche replied:
Why is the hashing needed? We just need to know whether it is a column name or not? 
Or you want to raise a different error message?

jorisvandenbossche replied:
And it certainly checks hashing for some cases:

```
In [5]: {1} in pd.Index([1, 2, 3])
...
TypeError: unhashable type: 'set'
```

(just not an iter because as you said, it is hashable)

jorisvandenbossche reviewed (approved):
Looks good to me, just a minor comment, and the whatsnew note that Tom asked for.

h-vetinari reviewed (commented):
@jreback

h-vetinari committed:
Review (TomAugspurger)

h-vetinari committed:
Review (jreback)

h-vetinari reviewed (commented):
@jreback PTAL

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into set_index_custom

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -4150,10 +4145,26 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
+                    # for hashable types that are not keys;
+                    # treat as ValueError, not missing key
+                    tipo = type(col)
+                    raise ValueError(err_msg,
```
What is the reason for not treating it as a missing key?

h-vetinari replied:
Because at that point we're dealing with an object we have absolutely no idea about, and assuming it is a key is worse (see the `iter`-KeyError) than assuming it isn't one.

jorisvandenbossche replied:
No, I mean assuming it is a *missing* key. If it cannot be a key, it certainly is a missing key. Anyway, this is exactly the same behaviour as what `__getitem__` is also doing (or what 0.23.4 was doing), so I don't see the reason to deviate from that

jreback commented on doc/source/whatsnew/v0.24.1.rst:
```diff
@@ -74,12 +74,12 @@ Bug Fixes
 **Reshaping**
 
 - Bug in :meth:`DataFrame.groupby` with :class:`Grouper` when there is a time change (DST) and grouping frequency is ``'1d'`` (:issue:`24972`)
+- Fixed regression where custom hashable types could not be used as column keys in :meth:`DataFrame.set_index` anymore (:issue:`24969`)
```
remove the word anymore

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into set_index_custom

h-vetinari committed:
Review (jreback & jorisvandenbossche)

h-vetinari committed:
Revert last two commits

h-vetinari commented on pandas/core/frame.py:
```diff
@@ -4150,10 +4145,26 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
+            else:
+                # everything else gets tried as a key; see GH 24969
+                try:
+                    hash(col) and self[col]
```
I tried it, and this does not work for the `iter` case,

jorisvandenbossche replied:
What do you mean with "it doesn't work". An iter will give False, so it will be seen as a missing key.

h-vetinari replied:
Yes, but the tests - correctly, IMO - do not expect a `KeyError` for an `iter`, as I'm arguing also in the other comment.

h-vetinari reviewed (commented):
We can do better than pre-#22486. As I'm arguing in a comment here, I'm suggesting we just shouldn't bother trying to "fix" *unhashable* custom types.

For that reason, I've reverted back to the passing commit which did not have crazy code gymnastic to give reasonable errors for absurd inputs.

(just to put this in perspective: pre-#22486 gave bad errors for reasonable inputs *and* the same errors for unhashable custom types. So this is a strict improvement becaus it gets rid of the former while not touching the latter)

jorisvandenbossche commented:
@h-vetinari did you keep a test for a TypeError being raised when given unhashable key? (I don't see it in the diff, but I assume it was already there?)

h-vetinari committed:
Review (jorisvandenbossche)

h-vetinari commented:
> @jorisvandenbossche: @h-vetinari did you keep a test for a TypeError being raised when given unhashable key? (I don't see it in the diff, but I assume it was already there?)

It's being tested through `set`, but raises a `ValueError` there. This was the reason I added `hash(col)` originally, because `test_set_index_raise_on_type` trying `df.set_index({1, 2, 3})` will not raise a `TypeError` but a `KeyError`.

Your suggestion of adding the types for the missing keys is not so bad, but a bit complicated to implement, especially if there are several keys (and the code is called from different places).

In any case, I think it's cleaner to keep raising a `ValueError` for the iterator/generator cases, and use the `KeyError` otherwise. I've implemented that in the last commit and also reflected it in the tests. I also added a missing key test for the custom type for you.

h-vetinari commented on pandas/core/frame.py:
```diff
@@ -4151,10 +4146,28 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
+                # raise a KeyError
+                tipo = type(col)
+                raise ValueError(err_msg + ' Received column of '
+                                 'type {}'.format(tipo))
```
I added this extra branch to give a more sensible error for the iterator/generator cases. It's maybe worth noting that it would be easy to re-add the capability to consume list-likes here, because tuples now *always* enter the first branch.

jorisvandenbossche replied:
I am not fully sure we should use `is_list_like` here either. You can have custom objects that are hashable, but also iterable (like a tuple is also the combination). And once it is iterable, `is_list_like` will give True.

jorisvandenbossche replied:
And the line we draw is on hashable (at least, if it is not hashable, the indexing machinery simply doesn't work), not on iterable (tuples are iterable, strings are iterable)

I am not saying I would use it myself, but there are certainly reasonable examples of things you can put in an object dtype array that are iterable. One example is a shapely geometry object which are iterable (iterate through coordinates points; strictly spoken, they are currently not hashable, but that is something they plan to fix).

Given 0.24.1, I think we might want to release today. Are you fine with us adding some small changes here to get us to merge it during the day? (keeping the broad rationale of course)

h-vetinari replied:
@jorisvandenbossche 
Feel free to make any desired changes to this PR, my strong preference is no `KeyError` for iterables though.

> there are certainly reasonable examples of things you can put in an object dtype array that are iterable

Why should an object dtype array (which is mutable) be hashable?

My suggestion would be that you try to come up with a hashable/iterable example that would also be list-like (from the POV of `is_list_like`, which would likely exclude various custom types, unless they directly inherit from `abc.Iterable`), before you start changing things around, but feel free.

jorisvandenbossche replied:
> Why should an object dtype array (which is mutable) be hashable?

Sorry, I just generally meant object dtype array-like, like an Index or Series. So that it can make sense to put hashable but iterable objects in an Index or Series

jorisvandenbossche replied:
> My suggestion would be that you try to come up with a hashable/iterable example that would also be list-like

As I said, eg a shapely geometry. 
(anything that is iterable is regarded 'list-like' in the eyes of `is_list_like`)

h-vetinari replied:
OK, didn't see the shapely example. Maybe use something like `is_list_like(col) and not hasattr('__hash__')` for the last `elif`?

h-vetinari replied:
Didn't see your comments pop up, and edited my answer independently about the shaply case

In any case, **I can't work on this any more today**. Failures seem to be due to dict-views being hashable on `PY2`. That case could simply be removed from the parametrization, it was an extra mile I tried to go.

h-vetinari committed:
Fix hashable listlikes (review jorisvandenbossche)

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into set_index_custom

h-vetinari committed:
Stabilize repr of frozenset

TomAugspurger reviewed (approved):
Looks good, thanks @h-vetinari.

Anyone else have comments on the implementation?

jreback commented:
I would rather revert entirely to 0.23.4 and have these changes sit in master.

TomAugspurger commented:
What are the pertinent differences between the 0.23.4 implementation and this implementation?

> On Feb 1, 2019, at 06:57, Jeff Reback <notifications@github.com> wrote:
> 
> I would rather revert entirely to 0.23.4 and have these changes sit in master.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

WillAyd commented:
I also would prefer to just reverting to pre 0.23.4 behavior for the next release. IIUC a lot of the churn here started with #22486 which improved error messaging in some fringe cases. 

My point I was trying to make on Gitter yesterday is I feel like we are cramming in a lot of churn to the code base for edge cases; would rather slow down and target a later release to look back at these comprehensively than keep patching on a tight timeline

jorisvandenbossche commented:
Note that this PR is *mostly* doing that, reverting to 0.23.4 behaviour. 
We could also actually 'git revert' #22486, but I am not sure that is needed (it also included some welcome code clean-up that is not related to those behavioural changes, additional tests, ..).

But it is true that we could revert more. Eg specifically revert the full "validation of the passed values" block that was added in #22486 (while keeping the code clean-up and additional tests). 
Personally I think it would be nice to keep the better error *messages* (not necessarily the error *type*)

h-vetinari commented:
> @TomAugspurger: What are the pertinent differences between the 0.23.4 implementation and this implementation?

Comparing to 0.23.4 / pre-#22486, the whole inspection part below is new (which does nothing but raise better errors, rather than just plonking whatever into `frame[col]`)

https://github.com/pandas-dev/pandas/blob/794f61d4d77d075b37e40144ed3adbda669ace64/pandas/core/frame.py#L4149-L4181

The part where the values are then collected has been cleaned up, but basically only very minor changes.
https://github.com/pandas-dev/pandas/blob/794f61d4d77d075b37e40144ed3adbda669ace64/pandas/core/frame.py#L4198-L4216

Compare with (last [commit](https://github.com/pandas-dev/pandas/commits/0.23.x) of 23.4):
https://github.com/pandas-dev/pandas/blob/0409521665bd436a10aea7e06336066bf07ff057/pandas/core/frame.py#L3889-L3913

Using `set` in the following was added (part of solving #22484):
https://github.com/pandas-dev/pandas/blob/794f61d4d77d075b37e40144ed3adbda669ace64/pandas/core/frame.py#L4225-L4227

TomAugspurger commented:
Can you take a look at that Will? Any concerns?

> On Feb 1, 2019, at 10:57, h-vetinari <notifications@github.com> wrote:
> 
> @TomAugspurger: What are the pertinent differences between the 0.23.4 implementation and this implementation?
> 
> Comparing to 0.23.4 / pre-#22486, the whole inspection part below is new (which does nothing but raise better errors, rather than just plonking whatever into frame[col])
> 
> https://github.com/pandas-dev/pandas/blob/794f61d4d77d075b37e40144ed3adbda669ace64/pandas/core/frame.py#L4149-L4181
> 
> The part where the values are then collected has been cleaned up, but basically only very minor changes.
> https://github.com/pandas-dev/pandas/blob/794f61d4d77d075b37e40144ed3adbda669ace64/pandas/core/frame.py#L4198-L4216
> 
> Compare with (last commit of 23.4):
> https://github.com/pandas-dev/pandas/blob/0409521665bd436a10aea7e06336066bf07ff057/pandas/core/frame.py#L3889-L3913
> 
> Using set in the following was added (part of solving #22484):
> https://github.com/pandas-dev/pandas/blob/794f61d4d77d075b37e40144ed3adbda669ace64/pandas/core/frame.py#L4225-L4227
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

h-vetinari commented:
> @WillAyd: IIUC a lot of the churn here started with #22486 which improved error messaging in some fringe cases.

That custom types got excluded was indeed an oversight (cf. lack of testing), but the churn had to do with the orthogonal review requirement to allow all list-likes (against my goals for the PR, and in spite of warnings about the ambiguity of the tuple case). I pointed this out again in other discussions #22225 / #24046, which lead to removing those list-likes again with #24762.

IOW, I don't quite agree that the improved errors were "fringe", and I will reject blame for the fact that these PRs caused churn. Rather, I was proactively trying to rectify/improve the situation before a release - also with this PR.

OTOH, I agree with you about not supporting these custom cases, and would have preferred to deprecate them already for 0.24.1. However, the prevailing consensus is to fix the regression first, so I tried to achieve that, and to provide another (IMO better) option than just reverting.

WillAyd commented:
@h-vetinari want to be clear not blaming you or that PR for anything - this kind of stuff just happens :-)

WillAyd commented:
I still think Jeff's approach on how to move forward is the best in this situation. OK with whatever you ultimately decide though

TomAugspurger commented:
@WillAyd do you have a chance to take a look through 
https://github.com/pandas-dev/pandas/pull/24984#issuecomment-459791046,

and here's a diff of the `set_index` implementation prior to #22486 and this PR

```diff
diff --git a/a b/b
index 5db9299..ed97d58 100644
--- a/a
+++ b/b
@@ -1,7 +1,46 @@
         inplace = validate_bool_kwarg(inplace, 'inplace')
+
+        err_msg = ('The parameter "keys" may be a column key, one-dimensional '
+                   'array, or a list containing only valid column keys and '
+                   'one-dimensional arrays.')
+
         if not isinstance(keys, list):
             keys = [keys]
 
+        missing = []
+        for col in keys:
+            if (is_scalar(col) or isinstance(col, tuple)):
+                # if col is a valid column key, everything is fine
+                # tuples are always considered keys, never as list-likes
+                if col not in self:
+                    missing.append(col)
+            elif isinstance(col, (ABCIndexClass, ABCSeries,
+                                  np.ndarray, list)):
+                # arrays are fine as long as they are one-dimensional
+                if getattr(col, 'ndim', 1) > 1:
+                    raise ValueError(err_msg)
+            elif is_list_like(col) and not hasattr(col, '__len__'):
+                # various iterators/generators are hashable, but should not
+                # raise a KeyError; we identify them by their lack of __len__.
+                # hashable listlikes with __len__ get tested as keys below.
+                tipo = type(col)
+                raise TypeError(err_msg + ' Received column of '
+                                'type {}'.format(tipo))
+            else:
+                # everything else gets tried as a key; see GH 24969
+                try:
+                    found = col in self.columns
+                except TypeError:
+                    tipo = type(col)
+                    raise TypeError(err_msg + ' Received column of '
+                                    'type {}'.format(tipo))
+                else:
+                    if not found:
+                        missing.append(col)
+
+        if missing:
+            raise KeyError('{}'.format(missing))
+
         if inplace:
             frame = self
         else:
@@ -11,7 +50,7 @@
         names = []
         if append:
             names = [x for x in self.index.names]
-            if isinstance(self.index, MultiIndex):
+            if isinstance(self.index, ABCMultiIndex):
                 for i in range(self.index.nlevels):
                     arrays.append(self.index._get_level_values(i))
             else:
@@ -19,29 +58,23 @@
 
         to_remove = []
         for col in keys:
-            if isinstance(col, MultiIndex):
-                # append all but the last column so we don't have to modify
-                # the end of this loop
-                for n in range(col.nlevels - 1):
+            if isinstance(col, ABCMultiIndex):
+                for n in range(col.nlevels):
                     arrays.append(col._get_level_values(n))
-
-                level = col._get_level_values(col.nlevels - 1)
                 names.extend(col.names)
-            elif isinstance(col, Series):
-                level = col._values
-                names.append(col.name)
-            elif isinstance(col, Index):
-                level = col
+            elif isinstance(col, (ABCIndexClass, ABCSeries)):
+                # if Index then not MultiIndex (treated above)
+                arrays.append(col)
                 names.append(col.name)
-            elif isinstance(col, (list, np.ndarray, Index)):
-                level = col
+            elif isinstance(col, (list, np.ndarray)):
+                arrays.append(col)
                 names.append(None)
+            # from here, col can only be a column label
             else:
-                level = frame[col]._values
+                arrays.append(frame[col]._values)
                 names.append(col)
                 if drop:
                     to_remove.append(col)
-            arrays.append(level)
 
         index = ensure_index_from_sequences(arrays, names)
 
@@ -50,7 +83,8 @@
             raise ValueError('Index has duplicate keys: {dup}'.format(
                 dup=duplicates))
 
-        for c in to_remove:
+        # use set to handle duplicate column names gracefully in case of drop
+        for c in set(to_remove):
             del frame[c]
 
         # clear up memory usage
```

And as @h-vetinari says, a simple revert of #22486 may very-well turn up dependent issues.

WillAyd commented on pandas/core/frame.py:
```diff
@@ -4158,10 +4153,29 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
+                    raise TypeError(err_msg + ' Received column of '
+                                    'type {}'.format(tipo))
+                else:
+                    if not found:
```
Can this not be consolidated with the first conditional branch in any way?

h-vetinari replied:
Good point, that is possible.

WillAyd commented:
@TomAugspurger yes noted. So what is our summary of changes compared to pre-existing behavior? 

- Better error messaging for incorrect types 
- Allowing setting of one-dimensional values (? - thought we didn't want to do this for some reason)

Anything else?

TomAugspurger commented:
IIUC, there's also the fix for trying to drop a column twice.

```python
In [5]: df = pd.DataFrame({"A": [1, 2, 3]})

In [6]: df.set_index(['A', 'A'])
Out[6]:
Empty DataFrame
Columns: []
Index: [(1, 1), (2, 2), (3, 3)]
```

That previously raised a `ValueError` late in the `set_index` method.

> Allowing setting of one-dimensional values (? - thought we didn't want to do this for some reason)


Can you clarify with an example? This continues to raise a KeyError, but with a better error message.

```python
In [12]: df.set_index([1, 2, 3])
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-12-f1418a759f48> in <module>
----> 1 df.set_index([1, 2, 3])

~/sandbox/pandas/pandas/core/frame.py in set_index(self, keys, drop, append, inplace, verify_integrity)
   4157
   4158         if missing:
-> 4159             raise KeyError('{}'.format(missing))
   4160
   4161         if inplace:

KeyError: '[1, 2, 3]'

```

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into set_index_custom

h-vetinari committed:
Review (WillAyd)

h-vetinari commented:
> @WillAyd: So what is our summary of changes compared to pre-existing behavior?
'- Better error messaging for incorrect types
'- Allowing setting of one-dimensional values (? - thought we didn't want to do this for some reason)

The second aspect was discussed at length in #24046, where the key point was that passing arrays had been enabled since forever (and documented - albeit not very extensively - in all the docs that are still online, with examples).

> Anything else?

The only things I can think of:
* printing all missing keys in the `KeyError` instead of just the first
* not failing on `drop=True` with duplicate columns (`drop=False` used to always pass in that case)
* clearer docs of the array-situation with #24762 
* passing doctests

h-vetinari commented:
> > @WillAyd: Allowing setting of one-dimensional values (? - thought we didn't want to do this for some reason)

> @TomAugspurger: Can you clarify with an example? This continues to raise a KeyError, but with a better error message.

@TomAugspurger, careful here, this is the heart of the ambiguity of lists that @jreback so dislikes, that was the core of the discussion in #24046, and why I opened #24697. A plain list is interpreted as a list of keys, while a list within a list gets interpreted as an array. This has also been the case since forever.

From your example:

```
>>> df.set_index(np.array([1, 2, 3]))
   A
1  1
2  2
3  3
>>> df.set_index([[1, 2, 3]])
   A
1  1
2  2
3  3
```

h-vetinari committed:
Unambiguous KeyError message

WillAyd commented:
@TomAugspurger I think this discussion is a good indicator of why I continue to agree with @jreback on reverting changes here. I don't think we *as the core team* have a good grasp on what has actually changed, so how can we effectively communicate this outwards and expect end users to understand before the minor release tonight?

TomAugspurger commented:
@WillAyd I was trying to understand your comment in 
https://github.com/pandas-dev/pandas/pull/24984#issuecomment-459827879. Can you clarify that with an example?

jreback commented:
again this discussion is out of control. moving this to 0.24.2 as this is not a critical bug fix. 

These diffs are not simple and have had and likely will have unintented side effects. I would much rather pretty much full revert this (for 0.24.1) or just leave for 0.24.2 (this partial revert). and remove all of the churn. last minutes changes have proven historically to be a disaster.

TomAugspurger commented:
I think this is ready, but I don't think it's a blocker for 0.24.1.

Do we have concrete suggestions on the PR itself? Are we OK to merge to master with the 0.24.2 tag, after 0.24.1 is cut?

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into set_index_custom

h-vetinari commented:
@jorisvandenbossche 

Actually, this PR is now independent of the improvements that were reverted *in the gathering* of the relevant values. I'll post a separate PR so that the discussion between the two does not get mixed up.

This PR now simply raises a better `KeyError` in the case of multiple missing keys, and will raise a `TypeError` (rather than `KeyError`) for list-likes without a length (i.e. iterators/generators, which definitely *shouldn't* be keys even if hashable list-likes are allowed). In other words, two out of the three points of #22484, but without breaking custom types.

h-vetinari committed:
Remove redundant whatsnew

h-vetinari commented:
@WillAyd @jreback 
Friendly ping. :)

Diff should be much easier to digest now.

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into set_index_custom

h-vetinari commented:
Not sure it needs a whatsnew as the improved error was already announced for 0.24.0., but happy to add one if desired.

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -4120,6 +4120,38 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
+                # arrays are fine as long as they are one-dimensional
+                if getattr(col, 'ndim', 1) > 1:
+                    raise ValueError(err_msg)
+            elif is_list_like(col) and not hasattr(col, '__len__'):
```
An alterative (in light of the general comment I made), an alternative might be to only specifically check for `isinstance(col, collections.abc.Iterator`) ? That should catch the common cases you mentioned like map, iter, generator, without the risk of catching things we didn't mean to catch.

BTW, one example that still passes here is a range object (so `df.set_index(range(n))` raises a "KeyError: "None of [Int64Index([0, 1, 2], dtype='int64')] are in the [columns]"", because it is seen as a missing key. 
That might be another case for which to add a specific check here (`or isinstance(col, range)`). We could also consider actually allowing range objects (converting it to an array), but that discussion is maybe better for another issue :-)

jorisvandenbossche reviewed (commented):
Personally, I am still a bit wary of this "special case" check (`elif is_list_like(col) and not hasattr(col, '__len__'):`) just to have a different error *type* for iterators/generators. It complicates the code, and it's a deviation from the standard rule. 
In the end, for the user, I think the most important is to have a helpful error *message*, and it doesn't matter that much whether the error *type* is KeyError or TypeError.

And to clear, apart from that technical detail, big +1 on improving the error messages.

h-vetinari commented:
> [...] just to have a different error _type_ for iterators/generators.

I think this is a must, because the following is just *really* confusing. Taken from a comment above:
```
>>> import pandas as pd
>>> import numpy as np
>>>
>>> df = pd.DataFrame(np.random.randn(5, 5), columns=list('ABCDE'))
>>>
>>> # Series-constructor uses iterators as if it was an array
>>> my_iter = map(str, df.A)
>>> pd.Series(my_iter)
0     0.8812414814169731
1    -1.0789948340108944
2    -0.2049936946205116
3    0.35952380472779194
4     -0.529524992759882
>>>
>>> # so why not do the same for df.set_index, which *also* works with arrays?
>>> my_iter = map(str, df.A)  # had been consumed previously
>>> df.set_index(my_iter)
Traceback (most recent call last):
[...]
KeyError: "None of [Index(['0.8812414814169731', '-1.0789948340108944', '-0.2049936946205116',\n       '0.35952380472779194', '-0.529524992759882'],\n      dtype='object')] are in the [columns]"
```
> There's a couple things that are confusing here. How did an iterator get turned into an `Index` object, without any further user-input? And why is that `Index` being tested as a column key?!

I'd be fine with those cases actually *working* (which was the original approach with list-likes from #22486, that unfortunately also included tuples), but the least would be a better error.

>  an alternative might be to only specifically check for `isinstance(col, collections.abc.Iterator)`?

That's completely fine by me.

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into set_index_custom

h-vetinari committed:
Review (jorisvandenbossche)

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into set_index_custom

h-vetinari commented:
@jreback @WillAyd 
Ping

WillAyd commented:
Admittedly haven’t had time to revisit this just yet. Have a backlog of issues and PRs out there if you’d like to help review and resolve some would certainly be appreciated and would free up review time for contributors on this one. 

> On Feb 12, 2019, at 11:26 PM, h-vetinari <notifications@github.com> wrote:
> 
> @jreback <https://github.com/jreback> @WillAyd <https://github.com/WillAyd>
> Ping
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/pandas-dev/pandas/pull/24984#issuecomment-463088919>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAlOURBUvjXZZtCuqTQsJ1i8eLq4NcOkks5vM744gaJpZM4aWcih>.
>

jorisvandenbossche commented:
@WillAyd I don't think this PR is really controversial any more (a lot of the earlier discussion is not relevant anymore for the current scope; what this PR now basically does is just providing a better error message. There should be no change in behaviour (except for the error type (KeyError -> TypeError) for iterators; what I am still discussing with @h-vetinari). 
So just wanted to say that I don't think it's that critical that you have time again to look at it before going forward with this, if you are fine with that, but of course if you prefer to have a look, no problem, then we wait with it.

jorisvandenbossche commented:
@h-vetinari thanks for the updates

> I think this is a must, because the following is just _really_ confusing. Taken from a comment above:

I agree that the current behaviour is very confusing (although with this PR (but without the check for iterators) it would already be better, as the key is not yet coerced to an Index object when doing the KeyError, so you would get something like "KeyError: <list_iterator object .. >"). 
But again, I tried to make it very clear I was talking about the error *type* and not the *message*.  Let's do it with an example. When passing an iterator/generator, you want to have it like this:

```
>>> df.set_index(my_iter)
....
TypeError: The parameter "keys" may be a column key, one-dimensional array, or a list containing
    only valid column keys and one-dimensional arrays. Received a column of type 'map_iterator'.
```

What I meant is that we could also do:

```
>>> df.set_index(my_iter)
....
KeyError: The parameter "keys" may be a column key, one-dimensional array, or a list containing
    only valid column keys and one-dimensional arrays. Received a key of type 'map_iterator', which
    was not found as a column name.
```

I slightly edited the error message (but that's just a suggestion, not the point I want to make), but the main difference is the TypeError -> KeyError. I personally don't think a user would care about that, they simply want to get good feedback on why it is failing (and it is mainly the error message that is important then). 
But having it as a KeyError should simplify the code a bit, and keep it more consistent with things like `__contains__` (or how dicts in python work).

h-vetinari commented:
@WillAyd 
I'd love to contribute more, but am completely underwater ATM. A ping is about the most I could muster (because this will affect other PRs down the road).

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into set_index_custom

h-vetinari commented:
@jorisvandenbossche 

Now I understand where you're coming from. I have nothing against improving the error message per se.

My point was more that we're already raising `TypeError` for things like sets and other unhashable types, and that it would make more sense to me that iterators are considered as wrong input types too.

I mean, I don't care too much about the error type for iterators, but it's the more consistent option to me. If you still prefer a KeyError, I'll change. Maybe @jreback (re: open change request) has an opinion on this as well?

jreback commented on pandas/core/frame.py:
```diff
@@ -4103,6 +4103,37 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
+            elif isinstance(col, Iterator):
+                # various iterators/generators are hashable, but should not
+                # raise a KeyError; other list-likes get tested as keys below.
+                tipo = type(col)
```
just put type(col) directly rather than adding another line here

h-vetinari replied:
Done.

jreback commented on pandas/core/frame.py:
```diff
@@ -4103,6 +4103,37 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
+                try:
+                    found = col in self.columns
+                except TypeError:
+                    tipo = type(col)
```
same as above

jreback commented on pandas/core/frame.py:
```diff
@@ -4103,6 +4103,37 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
+                                    'type {}'.format(tipo))
+                else:
+                    if not found:
+                        missing.append(col)
```
why are you not just raising here? what's the point of collecting these?

h-vetinari replied:
The point is to show all missing keys, not just the first one (like `.loc` does as well). See also #22484.

jreback commented on pandas/core/frame.py:
```diff
@@ -4103,6 +4103,37 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
+        for col in keys:
+            if isinstance(col, (ABCIndexClass, ABCSeries, np.ndarray, list)):
+                # arrays are fine as long as they are one-dimensional
+                if getattr(col, 'ndim', 1) > 1:
```
!= 1, is a 0-dim numpy scalar valid?

h-vetinari replied:
Done

jreback commented on pandas/core/frame.py:
```diff
@@ -4103,6 +4103,37 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
+                # arrays are fine as long as they are one-dimensional
+                if getattr(col, 'ndim', 1) > 1:
+                    raise ValueError(err_msg)
+            elif isinstance(col, Iterator):
```
i think an iterator is actually legit,  you should convert to a list before all of the type checks at the start of the loop.

h-vetinari replied:
Ok, I'm now letting instances of Iterators pass through as array-likes (they get transformed to lists below).

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into set_index_custom

h-vetinari committed:
Review (jreback)

h-vetinari reviewed (commented):
@jreback 
Thanks for the review; PTAL

h-vetinari committed:
Retrigger after connectivity issues

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into set_index_custom

h-vetinari commented:
@jreback
All green again.

jorisvandenbossche reviewed (commented):
Can you add a test for a iterator of the wrong length?

h-vetinari committed:
Review (jorisvandenbossche)

h-vetinari commented:
> @jorisvandenbossche: Can you add a test for a iterator of the wrong length?

I just did. This was a good point, because the current code was raising inconsistently in different places depending on whether `append=False` resp. if the passed arrays had mixed length.

Added a separate check for that (otherwise `ensure_index_from_sequences` has no possibility of raising when getting an array of different length).

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into set_index_custom

h-vetinari committed:
move test for easier diff

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -21,6 +21,7 @@ Other Enhancements
 
 - :meth:`Timestamp.replace` now supports the ``fold`` argument to disambiguate DST transition times (:issue:`25017`)
 - :meth:`DataFrame.at_time` and :meth:`Series.at_time` now support :meth:`datetime.time` objects with timezones (:issue:`24043`)
+- :meth:`DataFrame.set_index` now works for instances of ``abc.Iterator``, provided their output is of the same length as the calling frame (:issue:`22484`)
```
use this issue number, what about the note for the closed issue in the PR itself? #24969

jreback replied:
does this still close #24969 ?

h-vetinari replied:
I updated the OP to clarify, and added the this PR as you asked

h-vetinari replied:
OK, I now see that #24969 is still open, but that was already closed by ...

jreback commented on pandas/tests/frame/test_alter_axes.py:
```diff
@@ -253,25 +256,47 @@ def test_set_index_raise_keys(self, frame_of_index_cols, drop, append):
+    @pytest.mark.parametrize('drop', [True, False])
+    def test_set_index_raise_on_len(self, frame_of_index_cols, box, length,
+                                    drop, append):
+        df = frame_of_index_cols  # has length 5
```
add the issue number as a comment

h-vetinari replied:
Used this PR as reference again

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into set_index_custom

h-vetinari committed:
Review (jreback)

h-vetinari reviewed (commented):
Thanks for the review, PTAL

h-vetinari commented:
One of the azure jobs failed with a `ResourceWarning` (they're back? yay!)

```
[...]
s...........sys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name=0 mode='r' encoding='UTF-8'>

=================================== FAILURES ===================================
______________ test_chunks_have_consistent_numerical_type[python] ______________
[gw0] linux -- Python 3.6.8 /home/vsts/miniconda3/envs/pandas-dev/bin/python

all_parsers = <pandas.tests.io.parser.conftest.PythonParser object at 0x7f37938b8b00>

    def test_chunks_have_consistent_numerical_type(all_parsers):
        parser = all_parsers
        integers = [str(i) for i in range(499999)]
        data = "a\n" + "\n".join(integers + ["1.0", "2.0"] + integers)
    
        # Coercions should work without warnings.
        with tm.assert_produces_warning(None):
>           result = parser.read_csv(StringIO(data))

pandas/tests/io/parser/test_common.py:1078: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <contextlib._GeneratorContextManager object at 0x7f37472facc0>
type = None, value = None, traceback = None

    def __exit__(self, type, value, traceback):
        if type is None:
            try:
>               next(self.gen)
E               AssertionError: Caused unexpected warning(s): [('ResourceWarning', ResourceWarning('unclosed <ssl.SSLSocket [closed] fd=18, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6>',), '/home/vsts/work/1/s/pandas/io/parsers.py', 2854)].
```

Could someone please restart it? @jreback @TomAugspurger @jorisvandenbossche @gfyoung

h-vetinari commented:
Thanks to whoever restarted the job. Unfortunately, it seems to be a non-transient break (which is good, at least from the point of hunting `ResourceWarnings`).

My operating assumption is that a new boto release broke something, but unfortunately I can't see the version in the build script. The `conda list` call comes *after* the environment is installed.

h-vetinari committed:
Add 'conda list' for azure/posix after activate 'pandas-dev'

h-vetinari commented:
Added the `conda list` call in #25377 and merged it in here.

h-vetinari commented:
Nevermind, this is passing again. Seems those `ResourceWarnings` will have to wait for another day.

jorisvandenbossche commented:
Since it is passing again, can you then remove the conda list changes here?

h-vetinari commented:
@jorisvandenbossche 
Feel free to revert the last commit yourself, otherwise I'll do it tonight (Europe).

h-vetinari committed:
Revert "Add 'conda list' for azure/posix after activate 'pandas-dev'"

This reverts commit 87bd0a6f7b9377b24b2a185f8a658f9c23f5beac.

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into set_index_custom

h-vetinari commented:
@jreback @jorisvandenbossche 
This should hopefully be done now.

h-vetinari committed:
Reflect change in docstring

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -22,6 +22,7 @@ Other Enhancements
 - Indexing of ``DataFrame`` and ``Series`` now accepts zerodim ``np.ndarray`` (:issue:`24919`)
 - :meth:`Timestamp.replace` now supports the ``fold`` argument to disambiguate DST transition times (:issue:`25017`)
 - :meth:`DataFrame.at_time` and :meth:`Series.at_time` now support :meth:`datetime.time` objects with timezones (:issue:`24043`)
+- :meth:`DataFrame.set_index` now works for instances of ``abc.Iterator``, provided their output is of the same length as the calling frame (:issue:`22484`, :issue:`24984`)
```
does this PR also re-enable custom classes? pls add a note for this as well, or is this just a code reorg?

h-vetinari replied:
@jreback:
Minimal reorg, just better errors and enabling iterators (due to your review).

The custom classes were re-enabled by #25085 (which took over the tests from this PR), which closed the regression #24969, and has a corresponding whatsnew note. I guess the issue didn't get closed yet, because I only noted that #25085 was an alternative to this PR (at the time, for solving #24969), but didn't add the "closes #24969" explicitly - sorry.

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into set_index_custom

h-vetinari commented:
@jreback 
Can we push this one over the finish line? (new commit was just due to whatsnew conflict)
